### PR TITLE
Update to matchms 0.20.X and include Spec2Vec similarity score

### LINKED ISSUE #275: matchms similarity: Add MetadataMatch
MetadataMatch - this one would add an additional input asking for the metadata field to match, in case of a numerical difference it has to include the tolerance. There could be a dropdown to select from possible fields like SMILES, INCHI, RETENTION_TIME, RETENTION_INDEX, FORMULA, PRECURSOR_MZ etc.

### LINKED ISSUE #277: matchms similarity: Add Spec2Vec module
Using Spec2Vec as a similarity metric by accepting a trained model as an additional parameter would be a cool feature!

For the procedure see [here](https://github.com/RECETOX/galaxytools/issues/276).

Relevant papers:
http://dx.doi.org/10.1371/journal.pcbi.1008724

Steps:
- [x] Train the model on provided testing data
- [ ] Include in matchms similarity wrapper
- [x] #315
- [ ] #316

Problems:
- [x] #290

### LINKED ISSUE #322: New matchms design
With the changes of the scores object, many new options to run computations with matchms are possible. Therefore, it would be great to update the matchms wrappers to reflect the changes in the package.

My suggestion would be to have the following tools:
* **matchms filtering** - take a single file as input and produce a single output
* **matchms formatting** - take a scores `.json` file and produce an excel sheet with all query x pair combinations as rows and each score layer being a column)
* matchms similarity - this should be split into 3 new tools, each taking as input either (1) a single file for symmetric input, (2) a query and reference file or (3) a scores object. All 3 new tools could share the same functionality for reading inputs.
  * **matchms metadata match** - use the metadata match similarity metric to produce a scores object based on metadata comparison and using the new sparse option to create a sparse layer which already filters out many pairs from the later spectral comparison (for example via retention index matching). The individual metadata fields could be pre-filled from a selector or the user could provide a free form input - I guess the selector is the safer choice for now, as this will integrate nicely with the metadata harmonization within matchms. The similarity score operates in two modes:
    * equal match - metadata is compared directly and has to be equal
    * tolerance - the score is included in the scores table if the metadata is within the tolerance
  * **matchms fingerprint similarity** - this can be run either in dense or sparse mode and can have a threshold so that it operates basically the same as the metadata match. This makes sense, since the fingerprint similarity computations have parallel implementations using numba, making the dense computation very efficient. A sparse option with threshold then allows using this as a mask to get for example only molecules which are similar in structure or so.
  * **matchms spectral similarity** - this is similar to the existing wrapper, offering the spectral similarity metrics and the individual parameters etc., using the sparse computation by default - I currently don't see a use case for the dense scores.
* **matchms filter spectral similarity by range** - this function only takes a score object and attempts to filter the last spectral similarity score that was calculated using a user provided threshold for the number of ion matches and the score to produce a filtered scores object.
* **matchms pipeline** - a new wrapper for running a pre-configured pipeline, taking as inputs a pipeline configuration file and the query and reference spectra and producing a scores object - this is more of a convenience tool which allows users to take their pipelines created within matchms to the cloud and simply execute them on a galaxy runner after iteratively creating their workflow.

### LINKED ISSUE #370: Add EDAM identifiers to matchms suite
Input data
- mass spectrum (0943)
- fingerprint (1365)

Format
- msp (missing)
- mgf (3651)
- json (3464)

Operations
- spectral library search (3801) (matchms spectral similarity)
- filtering (3695) (matchms filter)

Outputs
- structure similarity score (0888) (fingerprint similarity)
- distance matrix (2855) (all similarities -> maybe make more precise with spectral similarity score?)

Topics
- Metabolomics
- Compound libraries and screening (3343)

### PR DESCRIPTION
**Pull request by maximskorik**
* add an option to compute similarity using a Spec2Vec model
* added Spec2Vec dependency to the tool
* bumped matchms version to `0.20.0`
* bumped spec2vec version to `0.8.0`

- [x] add tests after https://github.com/bioconda/bioconda-recipes/pull/38659 and https://github.com/bioconda/bioconda-recipes/pull/38683 are published
- [x] update documentation (including scheme)
- [x] #370 

Closes #322 
Closes #277
Closes #275
Closes #370 

TODO:
- implement spec2vec similarity in own module

### CODE REVIEW

**Comment by bgruening** [comment]
See also https://github.com/conda-forge/staged-recipes/pull/21824

**Comment by hechth** [comment]
This PR also contains the new matchms pipeline design for #322

File: tools/matchms/matchms_formatter.xml [hunk 1]
```diff
@@ -25,56 +25,26 @@
...
             --sf "$scores" \
-            --o "$output" \
-            $method.selection \
-            #if $method.selection == "get-thresholded-data"
-            --st $method.scores_threshold \
-            --mt $method.matches_threshold
-            #else
-            --k $method.k
-            #end if
+            --o "$output"
```
Comment (bgruening) [inline tools/matchms/matchms_formatter.xml:28]: single quotes

File: tools/matchms/matchms_metadata_match.xml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        <requirement type="package" version="0.55.1">numba</requirement>
+        <requirement type="package" version="@TOOL_VERSION@">matchms</requirement>
+    </requirements>
+
+    <environment_variables>
+        <environment_variable name="MPLCONFIGDIR">/tmp</environment_variable>
+    </environment_variables>
+
+    <command detect_errors="exit_code"><![CDATA[
+        python3 ${python_wrapper}
```
Comment (bgruening) [inline tools/matchms/matchms_metadata_match.xml:20]: single-quotes please

File: tools/matchms/matchms_metadata_match.xml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    <expand macro="creator"/>
+
+    <requirements>
+        <requirement type="package" version="1.1.4">pandas</requirement>
+        <requirement type="package" version="0.55.1">numba</requirement>
+        <requirement type="package" version="@TOOL_VERSION@">matchms</requirement>
+    </requirements>
+
+    <environment_variables>
+        <environment_variable name="MPLCONFIGDIR">/tmp</environment_variable>
```
Comment (bgruening) [inline tools/matchms/matchms_metadata_match.xml:16]: please do not do that, use $TEMP or `\$_GALAXY_JOB_TMP_DIR` instead, so an admin can manipulate that setting from outside

File: tools/matchms/matchms_metadata_match.xml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            <when value="equal_match">
+                <param label="Metadata Match" name="metadata_match_field" type="select" display="radio"
+                    help="Metadata field to use for matching">
+                    <option value="smiles" selected="true">SMILES</option>
+                    <option value="inchi">InChI</option>
+                    <option value="inchi_key">InChIKey</option>
+                </param>
+            </when>
+        </conditional>
+    <param label="join type" name="join_type" type="select" display="radio"
```
Comment (bgruening) [inline tools/matchms/matchms_metadata_match.xml:73]: indentation is off here

File: tools/matchms/matchms_networking.xml [hunk 5]
```diff
@@ -22,6 +22,7 @@
     <configfiles>
         <configfile name="matchms_networking_cli">
             python3 ${__tool_directory__}/matchms_networking_wrapper.py \
+            --score_name "${score_layer_name}_score" \
```
Comment (bgruening) [inline tools/matchms/matchms_networking.xml:25]: single quotes please, for security

**Comment by xtrojak** [comment]
@hechth please include the figure depicting the relationships among individual `matchms`, wrappers as we discussed.

File: tools/matchms/matchms_fingerprint_similarity.xml [hunk 6]
```diff
@@ -0,0 +1,84 @@
+<tool id="matchms_fingerprint_similarity" name="matchms fingerprint similarity" version="@TOOL_VERSION@+galaxy0" profile="21.09">
+    <description>matchms fingerprint similarity calculation</description>
```
Comment (xtrojak) [inline tools/matchms/matchms_fingerprint_similarity.xml:2]: Can you provide a more explanatory description?

File: tools/matchms/matchms_metadata_match.xml [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            <output name="scores_out" value="metadata_match/inner_join_ri_20.json" ftype="json" />
+        </test>
+        <test>
+            <param name="use_scores" value="True"/>
+            <param name="scores_in" value="metadata_match/ri_match_60.json" ftype="json"/>
+            <param name="tolerance" value="20.0" />
+            <param name="join_type" value="left" />
+            <output name="scores_out" value="metadata_match/left_join_ri_20.json" ftype="json" />
+        </test>
+    </tests>
```
Comment (xtrojak) [inline tools/matchms/matchms_metadata_match.xml:122]: Missing test case for `equal match` option.

File: tools/matchms/matchms_networking.xml [hunk 8]
```diff
@@ -22,23 +22,29 @@
...
 
     <inputs>
         <param label="Similarity scores" name="scores" type="data" format="json"
-            help="matchMS similarity-scores file."/>
+            help="matchms similarity-scores file."/>
+        <param label="Score Layer" name="score_layer_name" type="select" help="Metric of the score layer to create a network from.">
+            <expand macro="similarity_metrics"/>
+        </param>
+        <param label="Identifier key" name="identifier_key" type="text"
+            help="Unique metadata identifier of each spectrum from which scores are computed."/>
```
Comment (xtrojak) [inline tools/matchms/matchms_networking.xml:47]: This parameter is optional, but running the wrapper without it seems to be always running into error:

```
Traceback (most recent call last):
  File "/m2b/home/trojak/temp/galaxytools/tools/matchms/matchms_networking_wrapper.py", line 38, in <module>
    main(argv=sys.argv[1:])
  File "/m2b/home/trojak/temp/galaxytools/tools/matchms/matchms_networking_wrapper.py", line 31, in main
    network.create_network(scores, args.score_name)
  File "/usr/local/lib/python3.11/site-packages/matchms/networking/SimilarityNetwork.py", line 116, in create_network
    msnet.add_nodes_from(unique_ids)
  File "/usr/local/lib/python3.11/site-packages/networkx/classes/graph.py", line 635, in add_nodes_from
    raise ValueError("None cannot be a node")
ValueError: None cannot be a node
```

with this warning:

```
WARNING:galaxy.model:Datatype class not found for extension 'gml'
```

**Review by xtrojak — CHANGES_REQUESTED** [review]
So much good work, thanks @hechth !

File: tools/matchms/matchms_metadata_match.xml (hunk 7 above)
Comment (hechth) [reply tools/matchms/matchms_metadata_match.xml:123]: solved in https://github.com/RECETOX/galaxytools/pull/321/commits/5c4c6864d054d50cedc0443772c9ff4359434d92

File: tools/matchms/matchms_fingerprint_similarity.xml (hunk 6 above)
Comment (hechth) [reply tools/matchms/matchms_fingerprint_similarity.xml]: I improved the description in https://github.com/RECETOX/galaxytools/pull/321/commits/97a56eb90529ee48fbae42a8dd338d47faca51df and also added an additional footnote with a link to the details about the used molecular fingerprint

File: tools/matchms/matchms_metadata_match.xml [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            <param name="references" value="similarity/rcx_gc-ei_ms_subset.msp" ftype="msp"/>
+            <param name="matching_type" value="equal_match" />
+            <param name="metadata_match_field" value="smiles" />
+            <output name="scores_out" value="metadata_match/smiles.json" ftype="json" />
+        </test>
+        <test>
+            <param name="use_scores" value="True"/>
+            <param name="scores_in" value="metadata_match/smiles.json" ftype="json"/>
+            <param name="matching_type" value="equal_match" />
+            <param name="metadata_match_field" value="inchi_key" />
```
Comment (xtrojak) [inline tools/matchms/matchms_metadata_match.xml:106]: CI failing on this.

```suggestion
            <param name="metadata_match_field" value="inchikey" />
```

**Comment by hechth** [comment]
> @hechth please include the figure depicting the relationships among individual `matchms`, wrappers as we discussed.

Done in https://github.com/RECETOX/galaxytools/pull/321/commits/0cd5d0d03f94f708abd713d2402534f621b353bf

File: tools/matchms/help.xml [hunk 10]
```diff
@@ -0,0 +1,58 @@
+<macros>
+    <token name="@HELP_matchms@">
+        Documentation
+            The python library matchms provides functions to convert, filter and compare mass spectrometry data.
+            For an overview of the different galaxy mnodules see the figure below
+            For detailed documentation on the tool see https://github.com/matchms/matchms/blob/master/README.rst
+            and https://matchms.readthedocs.io/en/latest/ for the python API.
+
+            To get more familiar with the library, there is a `tutorial`_ available which explains how to build a mass spectrometry data processing
+            pipeline with matchms.
```
Comment (xtrojak) [inline tools/matchms/help.xml:10]:
```suggestion
            The Python library matchms provides functions to convert, filter and compare mass spectrometry data.
            For an overview of the different galaxy modules, see the figure below.
            For detailed documentation on the tool, see https://github.com/matchms/matchms/blob/master/README.rst
            and https://matchms.readthedocs.io/en/latest/ for the Python API.

            To get more familiar with the library, there is a `tutorial`_ available which explains how to build a mass spectrometry data processing
            pipeline with matchms.
```

**Review by xtrojak — APPROVED** [review]

**Pull request merged**